From 56ded289100a2c89722ac2cb94a470b03bfb04ba Mon Sep 17 00:00:00 2001
From: Richard Liu <r66033@freescale.com>
Date: Mon, 18 Nov 2013 14:06:32 +0800
Subject: [PATCH 14/43] ENGR00286762 fix dual display swap rectange mouse
 cursor ghost

fixed by don't update layer visible region if it is not the latest display,
as all the display use one layer source, if layer's visable region updated
after PRIMARY display, the dirty region of EXTERNAL display will not correct.

enable swap rectange can decrease the bus loading for the case only part of
screen has update, below is bus loading result for mouse moving in settings.
without swap rectange: bus loading 83.1%
enable swap rectange: bus loading 43.1%

Signed-off-by: Richard Liu <r66033@freescale.com>

Conflicts:
	services/surfaceflinger/SurfaceFlinger.cpp

Conflicts:
	services/surfaceflinger/SurfaceFlinger.cpp
---
 services/surfaceflinger/DisplayDevice.cpp  |  4 +++-
 services/surfaceflinger/SurfaceFlinger.cpp | 32 ++++++++++++++++++++++++++++++
 services/surfaceflinger/SurfaceFlinger.h   |  4 ++++
 3 files changed, 39 insertions(+), 1 deletion(-)

diff --git a/services/surfaceflinger/DisplayDevice.cpp b/services/surfaceflinger/DisplayDevice.cpp
index 1544e0b..ce38d38 100644
--- a/services/surfaceflinger/DisplayDevice.cpp
+++ b/services/surfaceflinger/DisplayDevice.cpp
@@ -146,8 +146,10 @@ DisplayDevice::DisplayDevice(
     setProjection(DisplayState::eOrientationDefault, mViewport, mFrame);
 
 #ifdef EGL_ANDROID_swap_rectangle
+    // only enable swap rectange in PRIMARY and EXTERNAL display, hwcomposer
+    // has problem to support swap rectange in Virtual display.
     if (eglSetSwapRectangleANDROID(display, surface,
-            0, 0, mDisplayWidth, mDisplayHeight) == EGL_TRUE) {
+            0, 0, mDisplayWidth, mDisplayHeight) == EGL_TRUE && mType <= DISPLAY_EXTERNAL) {
         // This could fail if this extension is not supported by this
         // specific surface (of config)
         mFlags |= SWAP_RECTANGLE;
diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index 9231441..0215de5 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -1082,6 +1082,19 @@ void SurfaceFlinger::rebuildLayerStacks() {
         invalidateHwcGeometry();
 
         const LayerVector& layers(mDrawingState.layersSortedByZ);
+
+#ifdef EGL_ANDROID_swap_rectangle
+        bool forceUpdate = false;
+        const sp<DisplayDevice>& hw0(mDisplays[0]);
+        for (size_t dpy=1 ; dpy<mDisplays.size() ; dpy++) {
+            const sp<DisplayDevice>& hwi(mDisplays[dpy]);
+            if (hw0->getLayerStack() != hwi->getLayerStack()) {
+                 forceUpdate = true;
+                 break;
+            }
+        }
+#endif
+
         for (size_t dpy=0 ; dpy<mDisplays.size() ; dpy++) {
             Region opaqueRegion;
             Region dirtyRegion;
@@ -1090,8 +1103,17 @@ void SurfaceFlinger::rebuildLayerStacks() {
             const Transform& tr(hw->getTransform());
             const Rect bounds(hw->getBounds());
             if (hw->isDisplayOn()) {
+#ifdef EGL_ANDROID_swap_rectangle
+                // below change fix dual display swap_rectange mouse cursor ghost display issue.
+                // don't update layer visible region if it is not the latest display, as
+                // all the display use one layer source, if layer's visable region updated
+                // after PRIMARY display, the dirty region of EXTERNAL display will not correct.
+                SurfaceFlinger::computeVisibleRegions(layers,
+                        hw->getLayerStack(), dirtyRegion, opaqueRegion, (dpy == mDisplays.size() - 1) || forceUpdate);
+#else
                 SurfaceFlinger::computeVisibleRegions(layers,
                         hw->getLayerStack(), dirtyRegion, opaqueRegion);
+#endif
 
                 const size_t count = layers.size();
                 for (size_t i=0 ; i<count ; i++) {
@@ -1675,7 +1697,11 @@ void SurfaceFlinger::commitTransaction()
 
 void SurfaceFlinger::computeVisibleRegions(
         const LayerVector& currentLayers, uint32_t layerStack,
+#ifdef EGL_ANDROID_swap_rectangle
+        Region& outDirtyRegion, Region& outOpaqueRegion, bool updateLayerRegion)
+#else
         Region& outDirtyRegion, Region& outOpaqueRegion)
+#endif
 {
     ATRACE_CALL();
 
@@ -1774,6 +1800,9 @@ void SurfaceFlinger::computeVisibleRegions(
             dirty = visibleRegion;
             // as well, as the old visible region
             dirty.orSelf(layer->visibleRegion);
+#ifdef EGL_ANDROID_swap_rectangle
+	    if (updateLayerRegion)
+#endif
             layer->contentDirty = false;
         } else {
             /* compute the exposed region:
@@ -1803,6 +1832,9 @@ void SurfaceFlinger::computeVisibleRegions(
         aboveOpaqueLayers.orSelf(opaqueRegion);
 
         // Store the visible region in screen space
+#ifdef EGL_ANDROID_swap_rectangle
+	if (updateLayerRegion)
+#endif
         layer->setVisibleRegion(visibleRegion);
         layer->setCoveredRegion(coveredRegion);
         layer->setVisibleNonTransparentRegion(
diff --git a/services/surfaceflinger/SurfaceFlinger.h b/services/surfaceflinger/SurfaceFlinger.h
index ed29269..f79dd28 100644
--- a/services/surfaceflinger/SurfaceFlinger.h
+++ b/services/surfaceflinger/SurfaceFlinger.h
@@ -375,7 +375,11 @@ private:
     void invalidateHwcGeometry();
     static void computeVisibleRegions(
             const LayerVector& currentLayers, uint32_t layerStack,
+#ifdef EGL_ANDROID_swap_rectangle
+            Region& dirtyRegion, Region& opaqueRegion, bool updateLayerRegion);
+#else
             Region& dirtyRegion, Region& opaqueRegion);
+#endif
 
     void preComposition();
     void postComposition();
-- 
1.8.0

